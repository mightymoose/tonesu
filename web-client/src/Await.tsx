import React, {
  ReactElement,
  useContext,
  createContext,
  FunctionComponent,
  PropsWithChildren,
  useEffect,
  useState,
} from "react";

interface PendingPromise {
  type: "pending";
}

interface ResolvedPromise {
  type: "resolved";
  result: unknown;
}

interface RejectedPromise {
  type: "rejected";
  result: unknown;
}

type PromiseStatus = PendingPromise | ResolvedPromise | RejectedPromise;

const AwaitContext = createContext<PromiseStatus>({ type: "pending" });

interface AwaitProps {
  promise: Promise<unknown>;
}

interface ThenProps {
  children: (result: unknown) => ReactElement;
}

interface CatchProps {
  children: (result: unknown) => ReactElement;
}

export const Await: FunctionComponent<PropsWithChildren<AwaitProps>> = ({
  children,
  promise,
}) => {
  const [promiseStatus, setPromiseStatus] = useState<PromiseStatus>({
    type: "pending",
  });

  useEffect(() => {
    let mounted: boolean = true;
    setPromiseStatus({ type: "pending" });
    promise
      .then(
        (result) => mounted && setPromiseStatus({ type: "resolved", result })
      )
      .catch(
        (result) => mounted && setPromiseStatus({ type: "rejected", result })
      );

    return function () {
      mounted = false;
    };
  }, [promise]);

  return (
    <AwaitContext.Provider value={promiseStatus}>
      {children}
    </AwaitContext.Provider>
  );
};

export const Pending: FunctionComponent = ({ children }) => {
  const promiseStatus: PromiseStatus = useContext(AwaitContext);

  return promiseStatus.type === "pending" ? <>{children}</> : null;
};

export const Then: FunctionComponent<ThenProps> = ({ children }: ThenProps) => {
  const promiseStatus: PromiseStatus = useContext(AwaitContext);

  return promiseStatus.type === "resolved"
    ? children(promiseStatus.result)
    : null;
};

export const Catch: FunctionComponent<CatchProps> = ({
  children,
}: CatchProps) => {
  const promiseStatus: PromiseStatus = useContext(AwaitContext);

  return promiseStatus.type === "rejected"
    ? children(promiseStatus.result)
    : null;
};
