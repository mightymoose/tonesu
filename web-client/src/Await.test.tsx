import { render, screen } from "@testing-library/react";
import React from "react";
import { Await, Pending, Then, Catch } from "./Await";

let resolve, reject;

beforeEach(() => {
  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });

  render(
    <div data-testid="test-yoke">
      <Await promise={promise}>
        <Pending>Pending</Pending>
        <Then>{(result) => <h1>Resolved {result}</h1>}</Then>
        <Catch>{(result) => <h1>Rejected {result}</h1>}</Catch>
      </Await>
    </div>
  );
});

it("displays the pending state when the promise is pending", () => {
  const testYoke = screen.getByTestId("test-yoke");
  expect(testYoke).toMatchSnapshot();
});

it("displays the resolved state when the has resolved", async () => {
  resolve("Then");
  await screen.findByRole("heading");
  const testYoke = screen.getByTestId("test-yoke");
  expect(testYoke).toMatchSnapshot();
});

it("displays the rejected state when the has been rejected", async () => {
  reject("Reject");
  await screen.findByRole("heading");
  const testYoke = screen.getByTestId("test-yoke");
  expect(testYoke).toMatchSnapshot();
});
