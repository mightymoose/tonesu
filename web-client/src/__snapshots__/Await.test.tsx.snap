// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the pending state when the promise is pending 1`] = `
<div
  data-testid="test-yoke"
>
  Pending
</div>
`;

exports[`displays the rejected state when the has been rejected 1`] = `
<div
  data-testid="test-yoke"
>
  <h1>
    Rejected 
    Reject
  </h1>
</div>
`;

exports[`displays the resolved state when the has resolved 1`] = `
<div
  data-testid="test-yoke"
>
  <h1>
    Resolved 
    Then
  </h1>
</div>
`;
